### PR TITLE
[ML] Hyperparameter importance

### LINKED ISSUE #1594: [ML] Calculate hyperparameter importance
Bayesian optimization method defines a Gaussian process for predicting model accuracy from the hyperparameters. We can use functional ANOVA decomposition of this GP to obtain the hyperparameter importance values very similar to those of the total feature importance.

### PR DESCRIPTION
**Pull request by valeriy42**
This PR implements hyperparameter importance calculation using ANOVA decomposition of the Gaussian process model used in the Bayesian optimization routine. The results is recorded in the model metadata structure. For convenience, I record the anova variance as an absolute value as well as a fraction of the total variation. This "relative importance" simplifies the comparison between different parameters. 

Additionally, I record the best value for the individual hyperparameter s.t. it can be used in a follow-up job configuration and doesn't have to be looked up separately.

Example output:
```json
"hyperparameter_importance": [
        {
            "name": "downsample_factor",
            "value": 0.35355339059327386,
            "absolute_importance": 0.123,
            "relative_importance": 0.9123
        },
        {
            "name": "alpha",
            "value": 0.010766199582740555,
            "absolute_importance": 0.0123,
            "relative_importance": 0.0321
        }
    ]
```

In order to verify the analytical formulas, I implemented Sobol sequence random number generation, which leads to lower variance for Monte-Carlo estimations.

Closes #1594

### CODE REVIEW

File: lib/maths/CBayesianOptimisation.cc
Comment (tveasey) [inline lib/maths/CBayesianOptimisation.cc:302]: This will have rather bad cancellation errors if `m_KernelParameters(0)` is large and we know that the constant factor is proportional to `m_KernelParameters(0)^2`. Let's discuss numerics a bit offline.

Comment (tveasey) [reply lib/maths/CBayesianOptimisation.cc:302]: Also, I think we should be bounding variances to >= 0.0.

File: lib/maths/CBayesianOptimisation.cc
Comment (tveasey) [inline lib/maths/CBayesianOptimisation.cc:63]: Could we maybe think of a better name for these two functions, i.e. I think that these should reflect the fact they are coming from integrals of the kernel.

**Review by tveasey — COMMENTED** [review]
I've done a first pass through. Overall looks great. I've focused mainly on some robustness suggestions. The tests passing are pretty convincing that the implementation is right, but I think it might be worth seeing if we can tighten up the numerics in a couple of places. Let's talk about this a bit offline.

File: lib/maths/CBoostedTreeFactory.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeFactory.cc:186]: Good idea, fixed.

File: lib/maths/CBoostedTreeImpl.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc:1237]: I am getting sign-conversion warnings without the static cast, that's why I added it even if it looks ugly.

File: lib/maths/CSampling.cc
Comment (valeriy42) [reply lib/maths/CSampling.cc:836]: missed this one. Thank you.

File: lib/maths/CSampling.cc
Comment (valeriy42) [reply lib/maths/CSampling.cc:832]: fixed

File: lib/maths/CBayesianOptimisation.cc
Comment (valeriy42) [reply lib/maths/CBayesianOptimisation.cc:314]: Fixed

File: lib/maths/CBoostedTreeImpl.cc
Comment (droberts195) [reply lib/maths/CBoostedTreeImpl.cc:1237]: Could you just make `i` a `std::size_t` in its declaration in the `for` loop instead?  That's what we'd usually do unless there's a really pressing need to have a signed loop variable.

Comment (tveasey) [reply lib/maths/CBoostedTreeImpl.cc:1237]: > Could you just make i a std::size_t

I'd second this, but it be to suppress warnings Valeriy is seeing from `parameters` which is an Eigen vector and which expects integer indexes. I never see warnings on function argument type coercion. Also, I don't recall seeing them from our linux build. I wonder which compiler version this is?

File: lib/maths/CBayesianOptimisation.cc
Comment (valeriy42) [reply lib/maths/CBayesianOptimisation.cc:235]: Fixed.

File: lib/maths/CBoostedTreeImpl.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc:1237]: I am getting the `clang` wornings in the VS Code IDE. I'll change and check the gcc compiler workings.

File: lib/maths/CBayesianOptimisation.cc
Comment (valeriy42) [reply lib/maths/CBayesianOptimisation.cc:63]: Alright, I call them `integrate1dKernel` and `integrate1dKernelProduct`. Do you think it's better?

File: lib/maths/CBoostedTreeImpl.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc:1273]: Good point! I added the term in `estimateMemoryUsage` and implemented `initializeTunableHyperparameters` to (re-) populate the list after restoring the object. This way we don't have to change the schema. Please, let me know if I have forgot something.

File: lib/maths/CBayesianOptimisation.cc
Comment (valeriy42) [reply lib/maths/CBayesianOptimisation.cc]: I bounded the variance.

File: lib/maths/CBayesianOptimisation.cc
Comment (tveasey) [reply lib/maths/CBayesianOptimisation.cc]: Sounds good.

File: lib/maths/CBayesianOptimisation.cc
Comment (tveasey) [inline lib/maths/CBayesianOptimisation.cc:246]: As before, since we're differencing numbers proportional to theta_0^4 in some places I'm a little worried about cancellation errors. Using here as an example of what I have in mind (same approach applies elsewhere), I would propose introducing a scale. For example, the following changes to:
```
    double theta02{CTools::pow2(m_KernelParameters(0))};
    double scale{std::max(1.0, theta02)};
    double f0{this->anovaConstantFactor(Kinvf) / scale};
    double sum{0.0};
    input = (input - m_MinBoundary(dimension)) /
            (m_MaxBoundary(dimension) - m_MinBoundary(dimension));
    for (std::size_t i = 0; i < m_FunctionMeanValues.size(); ++i) {
        TVector x{this->transformTo01(m_FunctionMeanValues[i].first)};
        sum += Kinvf(static_cast<int>(i)) *
               std::exp(-(CTools::pow2(scaledKernelParameters[dimension]) +
                          MINIMUM_KERNEL_COORDINATE_DISTANCE_SCALE) *
                        CTools::pow2(input - x(dimension))) *
               prodXt(x, dimension);
    }
    return scale * (theta02 / scale * sum - f0);
```
I feel this doesn't complicate matters too much and deals with the edge case. Of course it only really matters if `|f0| >> |CTools::pow2(m_KernelParameters(0)) * sum - f0|`. Also these are approximations of the quantities we care about anyway and given we're using double precision we probably get away with it. I'll leave it to you to decide whether to roll out these sorts of change.

**Review by tveasey — COMMENTED** [review]
I did another pass. This is looking pretty close. The remaining things are mainly cosmetic. I also made a concrete suggestion on dealing with cancellation error.

**Comment by tveasey** [comment]
One last thing: failures of
```
lib.maths.CBoostedTreeTest.CBoostedTreeTest.testCategoricalRegressors
lib.maths.CBoostedTreeTest.CBoostedTreeTest.testBinomialLogisticRegression
lib.api.CDataFrameAnalyzerTrainingTest.CDataFrameAnalyzerTrainingTest.testRunBoostedTreeRegressionTrainingWithStateRecovery
```
indicate some error since these changes shouldn't affect any existing tests. The most likely cause is the refactor of hyperparameter handling. I didn't spot any obvious issue, but this needs resolving before merge.

File: include/maths/CBayesianOptimisation.h
Comment (valeriy42) [reply include/maths/CBayesianOptimisation.h:100]: Fixed

File: include/maths/CBoostedTreeUtils.h
Comment (valeriy42) [reply include/maths/CBoostedTreeUtils.h:45]: Fixed

File: include/maths/CBoostedTreeUtils.h
Comment (valeriy42) [reply include/maths/CBoostedTreeUtils.h:54]: fixed

File: lib/maths/CBoostedTreeFactory.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeFactory.cc:251]: Fixed

File: lib/maths/CBoostedTreeImpl.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc:1339]: removed

File: lib/maths/CBoostedTreeImpl.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc:1272]: removed

File: lib/maths/CBoostedTreeImpl.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeImpl.cc:1468]: removed

File: lib/maths/CSampling.cc
Comment (valeriy42) [reply lib/maths/CSampling.cc:833]: Missed this one. Thanks!

File: lib/maths/CBoostedTreeFactory.cc
Comment (valeriy42) [reply lib/maths/CBoostedTreeFactory.cc:87]: Removed

File: lib/maths/CBayesianOptimisation.cc
Comment (valeriy42) [reply lib/maths/CBayesianOptimisation.cc:246]: Better safe than sorry. I added your suggestion to the relevant functions.

**Comment by valeriy42** [comment]
Thank you for the review @tveasey . I addressed all your comments. It would be great if you could take another look. 

In the mean time I'll look at how to fix the unit test failures.

**Review by tveasey — APPROVED** [review]
Thanks for iterating on this. Good stuff! This LGTM once it passes CI.

**Comment by valeriy42** [comment]
retest

**Pull request merged**
